src/ethereum/EthereumEngine.ts

```diff
@@ -1637,6 +1687,93 @@ export class EthereumEngine extends CurrencyEngine<
 
     await super.saveTx(edgeTransaction)
   }
+
+  // ****************************************************************************
+  // Private methods
+  // ****************************************************************************
+
+  /**
+   * Finds one decoy address if configured and needed.
+   * This method is called periodically by addToLoop. It finds one address
+   * per invocation and lets the loop handle repetition.
+   */
+  async findDecoyAddress(
+    decoyAddressConfig: DecoyAddressConfig
+  ): Promise<void> {
+    const currentCount = this.decoyAddressCount
+    if (currentCount >= decoyAddressConfig.count) {
+      // Already have enough decoy addresses; make this task a no-op
+      return
+    }
+
+    // Get current block height
+    const blockHeightUpdate = await this.ethNetwork.check('fetchBlockheight')
+    if (blockHeightUpdate.blockHeight == null) {
+      this.warn('Cannot generate decoy address: failed to get block height')
+      return
+    }
+    const currentBlockHeight = blockHeightUpdate.blockHeight
+
+    // Build exclude list: own address + existing decoy addresses
+    const excludeAddressesSet = new Set(
+      this.subscribedAddresses.map(sa => normalizeAddress(sa.address))
+    )
+
+    // Generate one decoy address
+    const decoyAddress = await this.ethNetwork.findOneDecoyAddress(
+      decoyAddressConfig,
+      currentBlockHeight,
+      excludeAddressesSet
+    )
+
+    if (decoyAddress != null) {
+      if (
+        this.subscribedAddresses.every(
+          a =>
+            normalizeAddress(a.address) !==
+            normalizeAddress(decoyAddress.address)
+        )
+      ) {
+        this.subscribedAddresses.push(decoyAddress)
+      }
+
+      this.log(`Generated decoy address: ${decoyAddress.address}`)
+
+      // Ensure that the real address is in the list
+      if (
+        this.subscribedAddresses.every(
+          a =>
+            normalizeAddress(a.address) !==
+            normalizeAddress(this.walletLocalData.publicKey)
+        )
+      ) {
+        this.subscribedAddresses.push({
+          address: this.walletLocalData.publicKey,
+          checkpoint: this.walletLocalData.highestTxBlockHeight.toString()
+        })
+      }
+
+      // If the number of decoy addresses meet the config requirements,
+      // then we resubscribe our addresses to the core. This means the engine
+      // has to successfully finish getting all decoy addresses before the core
+      // will persist them.
```

Comment:
This seems problematic. If a user isn't logged in long enough we'll throw out all the generated decoys.